flecsi/execution/hpx/task_add_dependencies.h

```diff
@@ -0,0 +1,202 @@
+/*
+    @@@@@@@@  @@           @@@@@@   @@@@@@@@ @@
+   /@@/////  /@@          @@////@@ @@////// /@@
+   /@@       /@@  @@@@@  @@    // /@@       /@@
+   /@@@@@@@  /@@ @@///@@/@@       /@@@@@@@@@/@@
+   /@@////   /@@/@@@@@@@/@@       ////////@@/@@
+   /@@       /@@/@@//// //@@    @@       /@@/@@
+   /@@       @@@//@@@@@@ //@@@@@@  @@@@@@@@ /@@
+   //       ///  //////   //////  ////////  //
+
+   Copyright (c) 2020, Los Alamos National Security, LLC
+   All rights reserved.
+                                                                              */
+#pragma once
+
+/*! @file */
+
+#include <hpx/include/lcos.hpp>
+
+#include <flecsi/data/common/data_reference.h>
+#include <flecsi/data/dense_accessor.h>
+#include <flecsi/data/global_accessor.h>
+#include <flecsi/data/ragged_accessor.h>
+#include <flecsi/data/ragged_mutator.h>
+#include <flecsi/data/sparse_accessor.h>
+#include <flecsi/data/sparse_mutator.h>
+
+/*!
+ @file
+ @date Initial file creation: November 16, 2020
+ */
+
+#include <cstdint>
+#include <string>
+#include <vector>
+
+#include "mpi.h"
+
+#include <flecsi/coloring/dcrs_utils.h>
+#include <flecsi/data/data.h>
+#include <flecsi/data/dense_accessor.h>
+#include <flecsi/execution/context.h>
+
+#include "flecsi/utils/mpi_type_traits.h"
+#include <flecsi/utils/tuple_walker.h>
+#include <flecsi/utils/type_traits.h>
+
+namespace flecsi {
+namespace execution {
+
+/*!
+ The task_add_dependencies_t type can be called to walk the task args before
+ the task has run. This allows to ensure task dependencies be added to the
+ execution flow.
+
+ @ingroup execution
+ */
+
+struct task_add_dependencies_t
+  : public flecsi::utils::tuple_walker_u<task_add_dependencies_t> {
+
+  /*!
+   Construct a task_add_dependencies_t instance.
+   */
+
+  task_add_dependencies_t()
+    : has_dependencies(false), promise(), future(promise.get_future()) {}
+
+  /*!
+   FIXME: Need a description.
+
+   @tparam T                     The data type referenced by the handle.
+   @tparam EXCLUSIVE_PERMISSIONS The permissions required on the exclusive
+                                 indices of the index partition.
+   @tparam SHARED_PERMISSIONS    The permissions required on the shared
+                                 indices of the index partition.
+   @tparam GHOST_PERMISSIONS     The permissions required on the ghost
+                                 indices of the index partition.
+
+   */
+
+  template<typename T,
+    size_t EXCLUSIVE_PERMISSIONS,
+    size_t SHARED_PERMISSIONS,
+    size_t GHOST_PERMISSIONS>
+  void handle(dense_accessor<T,
+    EXCLUSIVE_PERMISSIONS,
+    SHARED_PERMISSIONS,
+    GHOST_PERMISSIONS> & a) {
+
+    auto & h = a.handle;
+    // Skip Read Only handles
+    if constexpr((SHARED_PERMISSIONS == ro) || (GHOST_PERMISSIONS == rw) ||
```

Comment:
Do the `EXCLUSIVE_PERMISSIONS` not need to be used to setup task dependencies?

Reply:
Excellent question. I essentially copied the conditions from the MPI backend, so I might have missed things. Is there an explanation of how (and which) permissions are to be interpreted to deduce dependencies? Or in general, what is encoded by the different types of permissions?

Reply:
The MPI backend does not do dependency analysis. The check on permission is only for ghost copy purpose. Thus the task_[epi|pro]logue would skip handles that does not write to ghost or shared region.

For a more completed dependency analysis, the permission on Exclusive should be taken into account.

Reply:
Thanks for this explanation. I'm still unsure about the role of the various permissions, though. Would you be able to recommend some documentation or similar to read up about this?

Reply:
Minor correction. `GHOST_PERMISSION == rw` is used to signify the values of ghost cells are computed locally (flecsale does that) and thus ghost exchange is not necessary. Again, I don't know what the implication to dependency analysis is.

Reply:
https://github.com/laristra/flecsi/blob/1/flecsi/data/data.md#permissions